### PR TITLE
Speedup proptests for Chain struct in zebra-state

### LINKED ISSUE #1112: Speedup proptests for `Chain` struct in `zebra-state`
**Is your feature request related to a problem? Please describe.**

Right now the proptest Arbitrary impls we use to generate arbitrary chains for testing the `Chain` type are particularly slow. To work around this we currently only run one iteration of the proptest, but this clearly defeats the purpose of proptesting, so right now we essentially just have a couple of unit tests.

**Describe the solution you'd like**

There may be some ways that we can speed up the generation of arbitrary chains but in this particular case I don't think it should be our focus. For this test we care about proptesting the indexes we use when manipulating the Chain type, how many times we `push` or `pop_root` to show that for any combination of pushes or pops if they balance we end up with identical types. We should try to see if we can adjust the proptest such that it doesn't attempt to shrink the ValueTree used to generate the Chain itself, and instead only do searching on the indexes used against the chain we generate.

**Describe alternatives you've considered**

* speeding up proptest chain generation in `zebra-chain`

**Additional context**

This problem was introduced in https://github.com/ZcashFoundation/zebra/pull/1093 as part of the implementation of https://zebra.zfnd.org/dev/rfcs/0005-state-updates.html

### PR DESCRIPTION
**Pull request by fanatid**
## Motivation

Improve `proptest` performance for `Chain` struct because currently only 1 case set in `ProptestConfig`.

## Solution

Instead of generating a new chain for each `proptest` case generate it only once and searching only on the indexes.

Without PR, 1 case:
```
test service::non_finalized_state::chain::tests::forked_equals_pushed ... ok
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 25 filtered out; finished in 3.94s
test service::non_finalized_state::chain::tests::finalized_equals_pushed ... ok
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 25 filtered out; finished in 4.26s
```

With PR, 32 cases:
```
test service::non_finalized_state::chain::tests::forked_equals_pushed ... ok
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 25 filtered out; finished in 8.64s
test service::non_finalized_state::chain::tests::finalized_equals_pushed ... ok
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 25 filtered out; finished in 7.45s
```

Tested on `i5-8250U`.

## Related Issues

Closes https://github.com/ZcashFoundation/zebra/issues/1112.

## Follow Up Work

Nope.

### CODE REVIEW

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 1]
```diff
@@ -486,16 +493,54 @@ mod tests {
...
+        }
+
+        fn complicate(&mut self) -> bool {
+            self.count.complicate()
+        }
+    }
+
+    #[derive(Debug, Default)]
+    struct PreparedChain {
+        chain: Mutex<Option<Arc<Vec<PreparedBlock>>>>,
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs]: We're about to replace all our sync::Mutexes with futures-aware Mutexes, but we'll want to keep this one (#1995)
 
```suggestion
        // the proptests are threaded (not async), so we want to use a threaded mutex here
        chain: std::sync::Mutex<Option<Arc<Vec<PreparedBlock>>>>,
```

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 2]
```diff
@@ -486,16 +493,54 @@ mod tests {
...
+    }
+
+    impl Strategy for PreparedChain {
+        type Tree = PreparedChainTree;
+        type Value = <PreparedChainTree as ValueTree>::Value;
+
+        fn new_tree(&self, runner: &mut TestRunner) -> NewTree<Self> {
+            let mut chain = self.chain.lock().unwrap();
+            if chain.is_none() {
+                let tip_height = NetworkUpgrade::Blossom
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs]: We probably want to use Canopy here, to be consistent with #1898 
```suggestion
                let tip_height = NetworkUpgrade::Canopy
```

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 3]
```diff
@@ -486,16 +493,54 @@ mod tests {
...
+                    .activation_height(Network::Mainnet)
+                    .unwrap();
+                let ledger_state = LedgerState::new(tip_height, Network::Mainnet);
+                let blocks = Block::partial_chain_strategy(ledger_state, 100)
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs]: Can we make the 100 into a constant?

I think it's something like…
```suggestion
                let blocks = Block::partial_chain_strategy(ledger_state, MAX_PARTIAL_CHAIN_BLOCKS)
```

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 4]
```diff
@@ -506,25 +551,22 @@ mod tests {
         proptest!(ProptestConfig::with_cases(env::var("PROPTEST_CASES")
                                           .ok()
                                           .and_then(|v| v.parse().ok())
-                                          .unwrap_or(1)),
-        |((chain, count) in arbitrary_chain_and_count())| {
-            let chain = chain.0;
-            let fork_tip_hash = chain[count - 1].hash();
+                                          .unwrap_or(32)),
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs]: Can we make this into a constant?
```suggestion
                                          .unwrap_or(DEFAULT_PARTIAL_CHAIN_PROPTEST_CASES)),
```

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 5]
```diff
@@ -537,27 +579,24 @@ mod tests {
         proptest!(ProptestConfig::with_cases(env::var("PROPTEST_CASES")
                                           .ok()
                                           .and_then(|v| v.parse().ok())
-                                          .unwrap_or(1)),
-        |((chain, end_count) in arbitrary_chain_and_count())| {
-            let chain = chain.0;
+                                          .unwrap_or(32)),
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs]:
```suggestion
                                          .unwrap_or(DEFAULT_PARTIAL_CHAIN_PROPTEST_CASES)),
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks, this looks great!

I'm really happy to have better test coverage here. Even though the tests take slightly longer, they are still within our 10 second limit.

Just a few style issues and interactions with other changes - most of them are things we forgot to do as part of earlier changes.

File: zebra-state/src/service/non_finalized_state/chain.rs (hunk 4 above)
Comment (fanatid) [reply zebra-state/src/service/non_finalized_state/chain.rs]: Should I also change `PROPTEST_CASES` to const in other tests?
https://github.com/ZcashFoundation/zebra/blob/15843cfd6e046391672e1a9f91fb2bf81aecf3af/zebra-chain/src/work/tests/prop.rs#L50
https://github.com/ZcashFoundation/zebra/blob/15843cfd6e046391672e1a9f91fb2bf81aecf3af/zebra-chain/src/block/tests/prop.rs#L63

**Comment by fanatid** [comment]
hah, checked with `-p zebra-chain` locally instead `-p zebra-state`

File: zebra-state/src/service/non_finalized_state/chain.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/service/non_finalized_state/chain.rs]: Sure!

Making a named constant is helpful when there are multiple copies of the same number in a file, so they all get changed at the same time.

Comment (fanatid) [reply zebra-state/src/service/non_finalized_state/chain.rs]: Updated.

**Review by dconnolly — APPROVED** [review]
Love it! Waiting on @teor2345 's ✅

**Review by teor2345 — APPROVED** [review]
Looks good, thanks!

**Pull request merged**